### PR TITLE
Write your first comit-app tutorial

### PR DESCRIPTION
**Pull request by da-kami**
resolves: https://github.com/comit-network/comit.network/issues/4

Preview Link: https://deploy-preview-43--comit-network.netlify.app/docs/getting-started/comit-infrastructure

**Proposed Review Strategy:**

PLEASE no structural changes at this point (meaning moving stuff in different files or overall structure changes of the tutorial.)! 😅

There is always room for improvement, but if something is not a blocker (in the sense of it is wrong and HAS to be changed) lets please roll with the current version.
Happy to receive follow up PRs for structural changes if there is a strong feeling that they are necessary.

The current version of the tutorial is complete and logical in itself.

Please check:

* Spelling, Grammar
* Wording (e.g. rephrasing parts, ...)

---

**Details:**

Versioning:

The tutorial was added for the `next` version (in `docs` folder, which serves as "master") and the current latest version `0.15.2`. I think all links should be fine, I wrote the tutorial under the premise that it will be for `0.15.2`.

I added two separate commits for it.
Please review the files that are in `docs` and not in `versioned_docs`. If it is annoying that the `versioned_docs` were already added I can also remove the commit for the review. Please ping me. (I added it, so the preview is correct, otherwise one always has to switch to the `next` version...).

Concerning the tutorial:

It focuses on the maker first now and then on the taker. The actor initialisation is still done for both in one file, because I think that it is good to see, that we use the same environment for that, and it basically is the same.

Happy to talk about the overall structure instead of writing everything up here :)

Other changes:

I started splitting out "building blocks" that can be included in multiple sections of the documentation.
This is done by using the [`mdx` `import` feature](https://mdxjs.com/getting-started#imports).
I use this for:

* Global Disclaimers
* comit-scripts description
* create-comit-app description

Note that some of that was done quite fast, there is still room for improvement.

I created `docs/shared` folder and a `shared` folder in the tutorial for shared "blocks" - I would keep that convention for future documentation writing. 
If we structure the blocks nicely we can easily repeat ourselves by including a block.

### CODE REVIEW

File: docs/tutorials/write-a-comit-app.md [hunk 1]
```diff
@@ -1,7 +1,59 @@
...
+This is a typescript tutorial for creating your first COMIT-app. 
+In this tutorial we build a simple command-line application that uses the COMIT protocol to exectute an atomic swap locally on your machine between Bitcoin and Ethereum.
+
+## The story
+
+// TODO Describe a simple scenario that introduces the roles and what they want to do.
+
+Alice still has some Ether lying around from the early crypto days.
+Recently she took a look at Ethereum 2.0 she could not believe her eyes. (...)
+Scared of the Ethereum developments she is thinking about moving some of her Ether into Bitcoin.
```
Comment (D4nte) [inline docs/tutorials/write-a-comit-app.md]: I don't think this piece of opinion has its place here (or on any doc produced by CoBloX).

File: docs/tutorials/write-a-comit-app.md [hunk 2]
```diff
@@ -1,7 +1,59 @@
...
+
+Bob always wanted to run his own decentralised exchange.
```
Comment (D4nte) [inline docs/tutorials/write-a-comit-app.md]: I am not sure to understand the point of adding such a narrative.

**Comment by D4nte** [comment]
Do you know why netlify is not building?

File: docusaurus.config.js [hunk 3]
```diff
@@ -6,7 +6,14 @@ module.exports = {
   favicon: 'img/favicon.ico',
   organizationName: 'coblox', // Usually your GitHub org/user name.
   projectName: 'comit-network', // Usually your repo name.
+  plugins: ['@docusaurus/plugin-google-analytics', '@docusaurus/plugin-google-gtag'],
   themeConfig: {
+    googleAnalytics: {
```
Comment (D4nte) [inline docusaurus.config.js]: I guess this should not be part of this PR

**Comment by D4nte** [comment]
👍 for 50%.

**Comment by bonomat** [comment]
Hi @da-kami : before I can do the 50% PR review, can you summarize your core assumptions please? What is the goal of this tutorial, the key takeaways and what should we validate in the 50% review. Basically everything what was agreed on in the 10% review. 

In the 50% review neither the core assumptions nor the goal should be up for debate anymore, so it would be crucial to know them. The goal of the 50% review is to look at the overall structure and validate the direction again.

**Comment by da-kami** [comment]
> Hi @da-kami : before I can do the 50% PR review, can you summarize your core assumptions please? What is the goal of this tutorial, the key takeaways and what should we validate in the 50% review. Basically everything what was agreed on in the 10% review.
> 
> In the 50% review neither the core assumptions nor the goal should be up for debate anymore, so it would be crucial to know them. The goal of the 50% review is to look at the overall structure and validate the direction again.

Assumptions:
The developer has navigated through the comit.network page.
The developer already has tried running the [demo](https://comit.network/docs/getting-started/create-comit-app/) (might get a rename, see https://github.com/comit-network/comit.network/issues/49 )
The developer has already had a look at the dev-setup (this section does not exist yet, we described it here though: https://github.com/comit-network/comit.network/issues/49 ); parts of the Setup section of the tutorial might move there to avoid duplication.

Goal:
Make the developer understand:
- Roles of Taker and Maker and why there are 2 separate apps.
- Negotiation vs. Execution
- How to use the relevant classes of the SDK for coding.
- Create the btc_eth example of create-comit app together.

Note: Currently the tutorial does not match the example one-to-one. I did not introduce the `lib.ts` abstraction for creating the actors in the tutorial because I feel it adds unnecessary complexity. I would reflect on that later again.

Please validate that the tutorial is understandable, has a "red thread", consistent flow and all important points explained (e.g. mention of all the important SDK classes and what they are for). Once the tutorial is good we can decide if we want to change parts of the `btc_eth` example in the create-comit-app example code or adapt the tutorial.

**Comment by D4nte** [comment]
> We could add even another layer, depicting Negotiation and Execution as separate fold-down sub-section. What do you think @D4nte ?

The current structure looks good to me.

**Comment by D4nte** [comment]
👍 It looks good. I have not yet read in details as I am not sure the best way to move forward as there is some nitpicking I'd like to do `typescript` -> `TypeScript` .

Please note I think you should keep a summary of how the file look like at the end of each subsection. I don't think it's redundant.

**Comment by da-kami** [comment]
> 👍 It looks good. I have not yet read in details as I am not sure the best way to move forward as there is some nitpicking I'd like to do `typescript` -> `TypeScript`. Maybe I'll do a follow-up PR once merged?
> 
> Please note I think you should keep a summary of how the file look like at the end of each subsection. I don't think it's redundant.

I agree, with the changes that we introduced by splitting it up we have to do that. But keep in mind that it will be very annoying in maintaining this because the example keeps growing through the files. 
So if we change something in e.g. the initialisation stage, we have to pull it through to all other files.

**Comment by da-kami** [comment]
We run into problems when linking to the SDK.

We recently changed the `MakerNegotiator` and `TakerNegotiator` in the code. In the tutorial I link to the current SDK version of those classes.
 
These links will of course not be updated to the latest version once we actually do the SDK versioning ticket.

Underlying more general problem: Shall we link to SDK classes/interfaces from the tutorial? 
I think it does have value. But will be very hard to maintain. I don't see a smart way for links not becoming outdated this at the moment.

Shall I:

1. Leave the links and create a follow-up ticket?
2. Remove all the links to the SDK classes / interfaces?

@D4nte @bonomat

File: docs/tutorials/write-your-first-comit-app/04-2-taker-takes-order.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+yarn run v1.22.0
+$ ts-node ./src/maker.ts
+COMIT Maker app
+[Maker] Bitcoin balance: 10, Ether balance: 1000
+Maker's Negotiation Service is listening on localhost:2318.
+Waiting for someone to take my order at: http://127.0.0.1:2318
+✨  Done in 11.01s.
+```
+
+This is, because the maker was waiting for the incoming swap request on cnd and has received it from the taker.
```
Comment (da-kami) [inline docs/tutorials/write-your-first-comit-app/04-2-taker-takes-order.md]: @D4nte @bonomat With the current split in negotiation steps and execution steps we run into the  problem, that if the taker at this stage takes the order, an actual swap request is triggered through cnd. 
Both applications then terminate (because the execution was not added yet).
But the swap request of course still exists in cnd. 

Given the "simplified" swap-listening mechanism on the maker side, we always chose the first swap. 
On the taker side, however, we will always step through fetching an order and then initialising the swap from taking that order.
This means that after this stage the remaining swap in cnd will cause trouble.

I cannot just omit that detail, otherwise devs might run into that problem and get confused.
How shall I deal with this? I see three possible solutions:

1. Should we add matching logic, so it does not pick "the first" swap in the list?
2. Should we just let the tutorial user restart the environment?
3. Should we combine the "Taker takes order" section with the execution for the Taker? (probably the easiest, but I feel the separation between negotiation and execution is not that nice then.)

Comment (bonomat) [reply docs/tutorials/write-your-first-comit-app/04-2-taker-takes-order.md]: I'm not sure if I follow your problem. What do you mean with?

> Both applications then terminate (because the execution was not added yet).

Provided that this example is run one once, then I think everything is fine as 1 order is taken from the taker and then all other steps are executed. There is only 1 swap available on both sides at any time.

File: docs/tutorials/write-your-first-comit-app/04-2-taker-takes-order.md [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+It is important to understand, that the swap actually still exists in the maker cnd and the taker cnd!
+The two cnd nodes were started as part of the dev-environemnt and don't know about the taker and the maker app.
+So, both the maker's cnd node and the taker's cnd node are still waiting for the swap to be executed.
+
+Given our simple program logic, this is a problem now, because we are not detecting that there is a swap already in progress.
+The maker will always pick the first swap, while the taker will (once running through taking an order again) be on a different swap.
+
+> Note by Daniel: How should we solve this? The example is a bit incomplete here... 
+> Should we add matching logic, so it does not pick "the first" swap in the list?
```
Comment (bonomat) [inline docs/tutorials/write-your-first-comit-app/04-2-taker-takes-order.md]: I'm against this. The logic around this is not that straight forward. You would end up with a fully fledged maker.

File: docs/tutorials/write-your-first-comit-app/04-2-taker-takes-order.md [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+The maker's cnd node registered that and the maker app terminated when it noticed it.
+Since the logic for "waiting on incoming swaps" on the maker side is rather simple (it always choses the first one) the maker app will just terminate when started again (because there is already a swap in the cnd queue).
+
+
+It is important to understand, that the swap actually still exists in the maker cnd and the taker cnd!
+The two cnd nodes were started as part of the dev-environemnt and don't know about the taker and the maker app.
+So, both the maker's cnd node and the taker's cnd node are still waiting for the swap to be executed.
+
+Given our simple program logic, this is a problem now, because we are not detecting that there is a swap already in progress.
+The maker will always pick the first swap, while the taker will (once running through taking an order again) be on a different swap.
```
Comment (bonomat) [inline docs/tutorials/write-your-first-comit-app/04-2-taker-takes-order.md]: Unfortunately I can't follow your reasoning here.

File: docs/tutorials/write-your-first-comit-app/04-2-taker-takes-order.md [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+> Note by Daniel: How should we solve this? The example is a bit incomplete here... 
+> Should we add matching logic, so it does not pick "the first" swap in the list?
+> Should we just let the tutorial user restart the environment?
```
Comment (bonomat) [inline docs/tutorials/write-your-first-comit-app/04-2-taker-takes-order.md]: That's probably for the best. We can indicate that it's a one-shot environment and the developer needs to restart the environment if something did not work as expected. 
If the swap was executed successfully, the developer should be able to just restart the whole example.

File: docs/tutorials/write-your-first-comit-app/04-2-taker-takes-order.md [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+> Should we combine the "Taker takes order" section with the execution for the Taker? (probably the easiest, but I feel the separation between negotiation and execution is not that nice then.)
```
Comment (bonomat) [inline docs/tutorials/write-your-first-comit-app/04-2-taker-takes-order.md]: Yes, I would do so.

**Comment by D4nte** [comment]
> We run into problems when linking to the SDK.
> 
> We recently changed the `MakerNegotiator` and `TakerNegotiator` in the code. In the tutorial I link to the current SDK version of those classes.
> 
> These links will of course not be updated to the latest version once we actually do the SDK versioning ticket.
> 
> Underlying more general problem: Shall we link to SDK classes/interfaces from the tutorial?
> I think it does have value. But will be very hard to maintain. I don't see a smart way for links not becoming outdated at the moment.
> 
> Shall I:
> 
>     1. Leave the links and create a follow-up ticket?
> 
>     2. Remove all the links to the SDK classes / interfaces?
> 
> 
> (For the current SDK version these links are quite essential for the understanding, because e.g. the `Offer` exists once as interface used by maker and taker and once as class only relevant to the taker. Same for the `Negotiator` having the same name for maker and taker... We already fixed that by putting explicit names again but only in code, not in generated docs.)
> 
> @D4nte @bonomat

Once version are supported in the documentation I Would link to the version that is used to write the tutorial.

File: docs/tutorials/write-your-first-comit-app/04-2-taker-takes-order.md (hunk 4 above)
Comment (D4nte) [reply docs/tutorials/write-your-first-comit-app/04-2-taker-takes-order.md]: - Let the dev restart at the environment at this stage.
- Open an issue to fix that, this is a sign that we should provide a better API. The Maker only accepts the swap that the taker took. The Maker should know what swap he accepted via the negotiation interface and we should be manipulating this swap (`swapHandle`) instead of grabbing the first ongoing swap returned by cnd.

File: docs/tutorials/write-your-first-comit-app/03-2-maker-publishes-order.md [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+        }
+    });
+}
+```
+
+The maker uses the `ComitClient` to wait for incoming swaps.
+Once the taker takes the order he will trigger a swap request in his cnd that will be detected by the maker's cnd.
+
+> Notes by Daniel: This is a bit confusing. I feel the names should be harmonised (swap on the taker side, swap-handle on the maker)
+> Additionally: Is there any matching logic that checks if the swap matches the order on the maker side? (If so it is not obvious in the example...)
```
[the remark this answers is not in the document]
Comment (D4nte) [reply docs/tutorials/write-your-first-comit-app/03-2-maker-publishes-order.md]: Feel free to do patches on the SDK if there is anything that is confusing/do not make sense~

**Comment by D4nte** [comment]
> What I heavily dislike are the inline coding examples and splitting up 1 file over several passages.
> This is a maintenance hell and we should solve this somehow. Does docusaurus really not support including source code files?
> 
> We really should solve this before publishing this tutorial.

What is your ideal solution? The tutorial automatically updates when the code change in create-comit-app repo? How would this work? How is this maintenance hell?

The tutorial is fixed in time and it should be clear that it is a tutorial to build the example from create-comit-app@X.Y.Z with comit-sdk@A.B.C.

In term of maintenance, then yes, we could update the tutorial time to time, but only when we feels that there are such drastic changes in the comit-sdk interface that we want new dev to be aware of them.

**Comment by da-kami** [comment]
@bonomat @D4nte thanks for your input.

Going the comments I don't think we have consensus at the moment. 

@bonomat the original tutorial was actually in one file - one really long file. I discussed this with @D4nte and we took the decision to split it up. The idea for the original one-file version was not to have repeating code parts. The final code would have been displayed at the very bottom. This, however, is not as nice for the dev.

Yes, I feel that maintenance gets more because of splitting it up, I voiced it here:

> I agree, with the changes that we introduced by splitting it up we have to do that. But keep in mind that it will be very annoying in maintaining this because the example keeps growing through the files.
> So if we change something in e.g. the initialisation stage, we have to pull it through to all other files.

I do, however agree with @D4nte that the original file was really long. A split-up helps the dev, because the concerns are more clear and one can focus on specific sections.

Concerning complexity: It is questionable how much that can be reduced. Given, that we need actor setup, negotiation and execution in one example to showcase it that complexity cannot be much reduced. 

Here are some of my proposals towards a solution:

### Proposal 1 - take setup and actor initialisation out of the tutorial:

* Handle the environment setup and explanation outside of the tutorial in a separate section. Then take the knowledge about that for granted.
* Handle the actor setup completely separate outside the tutorial and have the `lib.ts` as resulting artefact. Then take the `lib.ts` for granted for the tutorial.

Advantages:

* Tutorial focusses on the interesting code parts

Disadvantages:

* The dev has to make sure he has all the pre-knowledge

### Proposal 2 - go back to one-file solution

* Take a step back and just do it on one file again (everything from setup till execution). See how devs react to it and then split up if necessary.

Advantages:

* The tutorial is a contained item that steps you from 0 to 100. There is (almost) no pre-knowledge needed.

Disadvantages:

* The tutorial is really long. Can be exhausting.

### Proposal 3 - stop here and postpone the tutorial

I feel, at the moment there are quite some unknowns involved in doing this tutorial good
Some of the problems we have are because there are still quite a lot of changes ongoing in e.g. SDK, the way cnd handles swaps (...).

We could take a step back and admit that it is too early to create this tutorial in a "perfect" way. 
Once we change the way cnd starts the swap execution with the new one-message protocols the SDK and then the tutorial will most likely change. 
When we iterate over the SDK and examples due to these changes some of the problem we currently have might just go away.

Additionally, once we did the versioning ticket we can tackle the link problem with more knowledge.

**Comment by bonomat** [comment]
Hi @da-kami : 

> @bonomat the original tutorial was actually in one file - one really long file. I discussed this with @D4nte and we took the decision to split it up. The idea for the original one-file version was not to have repeating code parts. The final code would have been displayed at the very bottom. This, however, is not as nice for the dev.

Sorry, I was not clear on that. I meant 1 file per section (see my structure proposal). In regards of the code: the code can be indeed split up over several passages but it should be sourced from one 1, e.g. like this: 

```
the `MakerNegotiator` can be initialized like this: 

https://github.com/comit-network/create-comit-app/blob/master/create/new_project/examples/btc_eth/src/maker.ts#L38-L59

Funding a swap like this: 

https://github.com/comit-network/create-comit-app/blob/master/create/new_project/examples/btc_eth/src/maker.ts#L120-L137

...


Full example, see here: 
https://github.com/comit-network/create-comit-app/blob/master/create/new_project/examples/btc_eth/src/maker.ts#L1-L181

```

**Comment by da-kami** [comment]
> Hi @da-kami :
> 
> > @bonomat the original tutorial was actually in one file - one really long file. I discussed this with @D4nte and we took the decision to split it up. The idea for the original one-file version was not to have repeating code parts. The final code would have been displayed at the very bottom. This, however, is not as nice for the dev.
> 
> Sorry, I was not clear on that. I meant 1 file per section (see my structure proposal). In regards of the code: the code can be indeed split up over several passages but it should be sourced from one 1, e.g. like this:
> 
> ```
> the `MakerNegotiator` can be initialized like this: 
> 
> https://github.com/comit-network/create-comit-app/blob/master/create/new_project/examples/btc_eth/src/maker.ts#L38-L59
> 
> Funding a swap like this: 
> 
> https://github.com/comit-network/create-comit-app/blob/master/create/new_project/examples/btc_eth/src/maker.ts#L120-L137
> 
> ...
> 
> 
> Full example, see here: 
> https://github.com/comit-network/create-comit-app/blob/master/create/new_project/examples/btc_eth/src/maker.ts#L1-L181
> ```

My idea of the tutorial was stepping the dev through it in a very guided way.
Additionally, the goal was that every stage should have a "working" version as result. 

We can move away from that and try it with the split you propose.
I think the biggest change would be the split by actors (focus on maker first then on taker).
I considered this in the beginning, but felt it would be nicer for the dev to have something he can "run" earlier. Maybe that assumption was not that good.

File: docs/tutorials/write-your-first-comit-app/01-setup.md [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+This tutorial is designed for Javascript/Typescript developers that want to integrate atomic swaps into their application.
+
+This section of the tutorial shows how to set up the project and prepare the maker and taker COMIT-apps.
+
+## Introduction
+
+The goal of the tutorial is to have a simple command-line application that handles the negotiation and execution of an atomic swap for both sides of a trade.
+
+For the negotiation phase we introduce **two roles**, the **maker** and the **taker**. The maker creates orders and publishes them. The taker takes orders published by the maker.
+In this tutorial our maker has Bitcoin and wants Ether. Our taker has Ether and wants Bitcoin.
```
[the remark this answers is not in the document]
Comment (da-kami) [reply docs/tutorials/write-your-first-comit-app/01-setup.md]: I changed the approach for parts of the section, resolving this for now.

**Comment by da-kami** [comment]
Somehow the preview is not correct, closing this PR and opening a new one.
